### ISSUE TITLE
[FEATURE] @Singleton or @Instance annotation

### ISSUE THREAD

**Issue by maywr** [projectlombok/lombok#3196]
## idea
the annotation would be needen in case of situations when you need only one instance of a class and ima show you what i mean

java:
```java
public final class Class extends Something {
  public static final Class INSTANCE = new Class()
}
```
lombok:
```java
@Singleton
public final class Class extends Something { }
 ```
 and also something like lazy feature would be cool (creating instance when `getInstance()` method is called)
 
 java:
 ```java
public final class Class extends Something {

  private static Class INSTANCE;

  public static Class getInstance() {
    if(INSTANCE == null) {
       INSTANCE = new Class();
    }
    return INSTANCE;
  } 
}
```
lombok: 
```java
@Singleton(lazy = true)
public final class Class extends Something { }
```
## why
that's a cool feature in my opinion and can be used in many cases when you have a class for some feature in programm

**Comment by janrieke** [comment]
The preferred idiom for singletons in Java is to use `enum`. This automatically ensures singularity and is inherently thread-safe.
One disadvantage is that you cannot lazily load it (because enums are always initialized on classload).

However, the "classic" way is unsafe due to reflection: Even if you make the constructor `private`, you can create new instances by making the constructor accessible and calling it.

I'm not a Lombok team member, but I'd vote against an annotation that generates a "classic" singleton implementation. One the other hand, an enum-based implementation is only one line of boilerplate (the `INSTANCE` line), making it not very useful.

**Comment by maywr** [comment]
enums can't extends classes, and the trouble you discovered is very easy to solve
```java
private Class() {
   if (INSTANCE != null) {
      throw new UnsupportedOperationException("you can't create instance of this class more than 1 time")
      }
}
```

**Comment by janrieke** [comment]
You may still mess with that by making the private `INSTANCE` accessible and setting it to `null`. Even final

But yes, I didn't think of the `extends` issue. That's at least a use case (don't know how frequent).

If Lombok would implement such an annotation, it must at least do the following:
- Make the class `static final`.
- Add the `INSTANCE` and the getter.
- Add a private constructor ensuring single instantiation, or alter the existing constructor to ensure this.
- Error out when there is more than one constructor.

PS: Your lazy getter check isn't thread-safe, but that can be fixed with double-checked locking. Even `final` won't help. But, well, that's really evil...

**Comment by cerebro84** [comment]
What if, as workaround, you used 

```java
@NoArgsConstructor(access = PRIVATE)
public final class SingletonClass extends Something {

 @Getter(lazy = true)
 private static final SingletonClass instance = new SingletonClass (); 
}
```

is not as convenient as your proposal, indeed it would be nice to have a `@Singleton` annotation in lombok, but in the meanwhile it might do the job.

**Comment by maywr** [comment]
yessss, that would work the same way but it requires more code than just `@Singleton` annotation

**Comment by rzwitserloot** [comment]
I'm not sure it's worth it. I'm pretty sure @rspilker and I discussed it a number of times and never considered it on the right side of the bang for the buck divide.

These are the use cases I'm aware of for it:

* You have a specific implementation of a general type, and this specific implementation is a singleton; either because they always are and the only reason we can't just use `@UtilityClass` is because there's a type hierarchy and `static` doesn't "do" hierarchy, or because the concept doesn't proscribe singleton-ness, but this take on it does. __The `enum` hack does not work here__, unless the type you want to have a singleton implementation for is an interface.
* Some combination of exotic factors just leads naturally to a situation where singleton is a good idea. By definition ignorable - not boilerplate.
* Misguided non-cases. Abuse of the pattern is rampant and is going to considerably mess up any attempts at studying use cases and dealing with feature requests:
    * A way to introduce global state. Most likely the author really wanted a dependency injection framework, or just wants to pass an instance around and needs to rethink their code flow design to avoid whatever made them go: "Oh no I do not want every third method in the entire project to grow a list of 18 arguments!" - or, in very rare cases, global state is warranted, but this is such a minefield that it feels immoral to even suggest that this is common enough to warrant a lombok feature. I think users of lombok should be able to intuit that anything lombok offers a feature for is boilerplate, and therefore acceptable in your average java project. Global state is not something I want lombok to 'peddle'. Hence, this is a non-case. It's a negative case, even: Rampant abuse of this pattern makes me want to veto `@Singleton`.
    * Singletons where it should just be utility methods. If it is right to introduce a global, unconfigurable single implementation as the only imaginable one - then why not just make a bunch of static methods? `java.lang.Math` does not have a singleton, and `Math.max` is a static method. Many usages of singleton out there should just be all-static methods instead, and `@UtilityClass` already exists to cater to this crowd. It's another negative use case: Singletons that should have just been utility classes is common, and merely suggesting this code style error is acceptable by introducing `@Singleton`, in addition to confusing matters by having both `@UtilityClass` and `@Singleton`, makes me want to veto `@Singleton`.
    * Singletons for things where alternate implementations really do exist - this is what dependency injection frameworks are all about. One impl of the twitter API for production use, but another for the testing infrastructure, that sort of thing. One way to get around dep injection frameworks is to have global state instead (the twitter API has a global 'go into test mode' flag or field and the test code `@Before`'s itself into test mode first, for example). I don't think this is right, and we shouldn't enable such bad code.
    * It could be `@UtilityClass` but the author wants to future-proof for the day that it cannot be. Author is wrong; making a singleton doesn't future-proof anything, as the vast majority of calls will just be of the form `SingletonType.getInstance().doThing()` which is equally useless to `SingletonType.doStaticThing()` as far as 'refactor it to an actual object you need to create and initialize first' purposes are concerned. Author is misguided and should just apply YAGNI and use `@UtilityClass`, or author correctly identified that future proofing is warranted, but should be adding a bunch of arguments, or rethinking the architecture somewhat to elevate this state to fields, or should possibly be using an injection framework instead and configuring it as singleton for now.
    * It could be `@UtilityClass` but author doesn't like the style or is misguided by the notion that 'static is bad' and hilariously thinks a singleton is the way out - in other words, fails to understand the underlying principle is 'global state is bad', or is letting style preferences beat objective, falsifiable damage to the quality of your codebase, either way, utterly misguided principle that makes them _think_ they want `@Singleton`, but, they really don't.


@janrieke 's `enum` suggestion is right out - he says that this is 'the preferred idiom'. I hope not, because if that is true, then the majority of the java ecosystem are misguided in this regard. The `enum` hack doesn't work for precisely the one time when singletons are definitely the right answer (when you need a specific impl of some class - enums cannot `extend` anything), and it is dirtying up your API: That a type is an enum is not and cannot be an implementation detail! You can't 'hide' enumness from an enum, it's right there, it will be `instanceof Enum`, the type hierarchy shows it to you, etcetera. A singleton has no relationship whatsoever to the concept of `enum`s, and the fact that your singleton does implement it is therefore merely accidental exposure of implementation details which is _bad_. Thus, doubly bad, thus, wrong. Given that it's so pervasive that someone whose programming skills I rate quite highly has been entrapped by this wildly wrong idea is making me want to write `@Singleton`, so this balances the scales out somewhat :)

There are two aspects that make writing singletons complicated:

* Laziness - the need to create the singleton object only when it is truly needed and not at any point before then. But this is __highly misguided__ and I think I ran into a scenario where laziness truly was needed __once__ in a very long programming career, compared to literally hundreds of blogposts, tutorials, book chapters, and questions about wanting singletons in a way that clearly indicates they (think they) need it. The java classloader system is __already lazy__ - a class is not loaded (and hence, not initialized, and hence, that singleton object is never made) until the class is used someplace. The time difference between when a class is initialized vs. when it is used is too small to matter, and the odds that the class must be loaded, but no instance is ever actually needed, is so tiny, that it's not worth writing code for. Especially considering that the vast majority of singletons out there in live java code and notably including the proposal here calls a no-args constructor which costs nothing. Worrying about laziness for this is __completely incorrect__. One of those mass delusion situations. You're nuts if you think you need this! Of course, if we ever do make a `@Singleton` and you write an actual constructor (or an initializing expression on an instance of course), we'll just leave it alone, and you can stick your pricey code in there, no problem. At any rate, if you want laziness and it's warranted (rare), then ensuring the initialization only ever occurs once in a thread safe fashion can be a little bit complicated.
* Serialization. In order to ensure only one instance exists even when you use java's own serialization mechanism can be a little bit tricky. Not overly so - we can trivially generate the right stuff here. It's unfortunate that we cannot _detect_ whether you need it, and given that java's own serialization in general is a big mistake, I'd have been great if we simply do not generate unless needed. We can generate it if you `extends` nothing (or literally `j.l.Object`, and `implements` nothing. But if you extends or implements anything, you might be serializable and therefore we have to generate the read/write replacement code. Yich: That code might trigger linter tools, doesn't do anything at all if you aren't serializing, and needlessly clutters up your private API (autocomplete dialogs within the class itself, for example). Also, the number of times the OpenJDK team is talking about serialization in a negative light + their prepensity to break stuff recently means I'm foreseeing unfortunate situations where we have to bend over backwards because OpenJDK messed with stuff.


----------------

There are solutions to all of these concerns:

* Have a `serialization` or whatnot parameter on the annotation. If it is unset:
    * If the `@Singleton` class doesn't extend anything and doesn't implement anything (or extends `j.l.Object`), don't generate it.
    * If the `@Singleton` class is declared to `implements` Serializable, generate it.
    * Otherwise, error out and demand that it is explicitly set to something.
* Have a wiki page that explains why you are highly likely to not need lazy singletons.
* Possibly add to the docs some notes on when singletons are likely not really what you want, couched in terms of 'this is not what `@Singleton` is for'.

But add all those caveats and is it even worth it?

**Comment by janrieke** [comment]
The recommendation to use `enum` comes from Joshua Bloch's _Effective Java_. One reason (among others) is the serialization issue you describe. Using `enum`, you simply don't have to worry about it, as `enum` serialization is handled by the JVM itself.

That being said: I never ever in my professional career had to manually implement a singleton. Whenever something like a singleton was necessary, I used a DI framework for it. So I'm also not convinced that an annotation for it is worth it.

**Comment by rzwitserloot** [comment]
> The recommendation to use enum comes from Joshua Bloch's Effective Java.

Appeal to authority is still a logical fallacy. But let's not throw Joshua under the bus here: If you have to __handroll__ the serialization replacement code, it's such a drag that I completely understanding looking for a way out and wanting it so badly that you're willing to accept some damage to your API - that you start considering that damage the lesser evil.

With lombok providing the feature, that reasoning is now moot and the API damage is no longer acceptable. Joshua may well agree with this sentiment - hence, the power that appeal to authority arguments do hold when we're specifically talking about code style is severely limited here.

I'd be on this like white on rice if there was a fairly easy way to use the baked in serialization+singleton-ness of java's `enum` feature without exposing the fact that you actually have an `enum` outside the package, but I don't think that's possible.

**Comment by ts14ic** [comment]
> Appeal to authority is still a logical fallacy

I don't think this is an appeal to authority, but rather an explanation of where did the `enum` thing come from. 
It truly is the "preferred idiom" (preferred as in statistics) by the virtue of a lot of people reading Effective Java and following its advice, as Joshua had a very convincing explanation of why `enums` be preferred, including security.
And not only that, but it's also short to write.

About enums not being able to extend classes. Any time I needed something like this, I actually used interfaces. 
And that was very rare. I mean, needing singletons is rare.

Which is to say, so-o-o rare, that I don't see why would anybody want an annotation so much for something like this. Is it going to save what... 2 lines of code across of a 100k lines of code project? 

And if that annotation is parameterized to have parameters for the method name, maybe field name, whether it's an eager singleton, or held in the static holder class, then the whole annotation becomes just as long, as implementing the said singleton by hand.

**Comment by rzwitserloot** [comment]
> Well, Lombok is also creating a mess of classes, where 50% of it is often implemented via annotations out of sight

That's an oxymoron (an argument that disproves itself, and is therefore trivially incorrect) - The whole point is - it is __out of sight__, an implementation detail. The problem with using `enum` to make singletons is that it's __in sight__. You can see it. It messes with your type hierarchies. You can't hide it.

> y out of sight, I mean, when you document your fields, and you simply don't see any of those @Data, @Getter, @Setter on top of the class.

That's __a good thing__, you're saying it like it's a bad thing. You document the what and perhaps the why. You never document the how, because that's an implementation detail. If you had been able to tell the difference between a hand-written setter and a lombok-generated one, that would be __really bad__, because it removes a large part of the very point of lombok: You can just have lombok generate that stuff for you, and if later you decide: Oh, wait a second, I need to do all sorts of non-boilerplate in there, you can just... do that. Without any user of your library being the wiser. It's entirely backwards compatible. __That is good__, not bad.

**Comment by ts14ic** [comment]
tl;dr: 
Ah, I see. 
We were just speaking about two different things.
Your concern for `enum` was focused around the type system and byte code, from what I understand.
My concern is focused on the effect on the source code semantics from developer perspective.

Full:

> The problem with using enum to make singletons is that it's in sight. You can see it. It messes with your type hierarchies. You can't hide it.

Yes, yes. The type hierarchy is messed up. This "type system perspective" we actually agree on.
I don't think anything was disproved though, because I was speaking about something else.
I was speaking about source code, not byte code.

As a developer, I don't care much, that using `enum` messes up my type hierarchy, because I never had a problem with it.
I use "interfaces" for polymorphism, and composition for code reuse. Inheriting implementation is my last resort, so I didn't have any need to extend an `enum` from other classes.

But I do care that when I see the keyword `enum`, I expect an enumeration of values.
Because that's what the keyword says.
Yet in case of `enum` singletons it's just one value instead of an enumeration, and not a mere value, but actually a fully fledged class.
This irks me. But it's a trade-off I can accept, because the benefits outweigh it for me.

> That's a good thing, you're saying it like it's a bad thing. 

Yes, from byte code perspective Lombok is doing a good job. When you use Lombok annotated classes from outside, they look just as if they were written by hand. It would be bad, if you'd require some special syntax to call Lombok accessors.

> Oh, wait a second, I need to do all sorts of non-boilerplate in there, you can just... do that

Yes, it is good, that you can trivially replace lombok generated code with hand-written code.

The bad thing I was talking about, was about what the source code of the class looks like, when you do that.
Or even when you don't do it.
Like, when I have a field marked `private`, but it's actually a public property. And this fact is out of my sight - my human sight - because the annotation is somewhere out of my screen space due to javadocs and what not.
(why would I mention javadocs in my previous message, if not for this perspective?)

And even if you don't have javadocs, sometimes you require additional invariants, or some normalization logic in getters and or setters, or withers, or builders.
And my eyes start glazing, because I now have to violently jump around the file to understand the classes intended semantics vs what the code says.
The "public interface" of the class becomes a blurry mess.
Where as in vanilla java, you have the private implementation details (fields) on one side, and all of its public API below. 

I do love me some `@EqualsAndHashCode` / `@ToString` though.
That thing is truly boring most of the time.
And absolutely trivial DTO classes, which can be annotated with some `@Data` and be done with it.

**Comment by ts14ic** [comment]
And the reason why I'm not as concerned about the type system, is because, uh... 
When I want to use some class, I don't see its public interface like a compiler does.
I open that class's file, and inspect its source code. And when Lombok is misused/overused, then I have to repeat the job of Lombok, by piecing together all of the annotations and hand rolled code in my head, just to understand what the cloud would've been without the annotations.

Maybe, someday somebody will develop an IDE plugin that shows some kind of preview of the resulting class :) .

To warp it all back into my original message about enums, and likening it to Lombok.
If somebody is willing to accept Lombok regardless of how it affects source code readability, then he's likely to accept what the `enum` singleton does to the source code readability as well.
Be it because he accepts the trade-off, or because he simply doesn't care.

**Comment by rzwitserloot** [comment]
Discussed it with @rspilker and we're not doing this. By a very large margin we think the use cases are rather disjunct, and very rare - not worth the maintenance burden. Just use `@UtilityClass` - in those rare cases you really need this, it probably isn't boiler plate. I know every patterns book and whatnot has a chapter on Singleton, but that's a red herring. You rarely need em, and if you think you need them a lot, you're probably doing it wrong.

**Comment by ts14ic** [comment]
> Just use `@UtilityClass`

I noticed you mentioned somewhere, that `@UtilityClass` is moving out of experimental soon. 
Is that true?
Curious, because the feature page says it won't move out, because of the non-star static import issue.
